src/validators/SessionKey/ERC20SessionKeyValidator.sol

```diff
@@ -0,0 +1,66 @@
+// SPDX-License-Identifier: MIT
+pragma solidity ^0.8.19;
+
+import { ISessionKeyValidationModule, UserOperation } from "./ISessionKeyValidationModule.sol";
+import { ECDSA } from "@openzeppelin/contracts/utils/cryptography/ECDSA.sol";
+import { MessageHashUtils } from "@openzeppelin/contracts/utils/cryptography/MessageHashUtils.sol";
+
+contract ERC20SessionValidationModule is ISessionKeyValidationModule {
+    /**
+     * @dev validates if the _op (UserOperation) matches the SessionKey permissions
+     * and that _op has been signed by this SessionKey
+     * @param _op User Operation to be validated.
+     * @param _userOpHash Hash of the User Operation to be validated.
+     * @param _sessionKeyData SessionKey data, that describes sessionKey permissions
+     * @param _sessionKeySignature Signature over the the _userOpHash.
+     * @return true if the _op is valid, false otherwise.
+     */
+    function validateSessionUserOp(
+        UserOperation calldata _op,
+        bytes32 _userOpHash,
+        bytes calldata _sessionKeyData,
+        bytes calldata _sessionKeySignature
+    )
+        external
+        view
+        returns (bool)
+    {
+        address sessionKey = address(bytes20(_sessionKeyData[0:20]));
+        // 20:40 is token address
+        address recipient = address(bytes20(_sessionKeyData[40:60]));
+        uint256 maxAmount = abi.decode(_sessionKeyData[60:92], (uint256));
+        {
+            address token = address(bytes20(_sessionKeyData[20:40]));
+
+            // we expect _op.callData to be `SmartAccount.executeCall(to, value, calldata)` calldata
+            (address tokenAddr, uint256 callValue,) = abi.decode(
+                _op.callData[4:], // skip selector
+                (address, uint256, bytes)
+            );
+            if (tokenAddr != token) {
+                revert("ERC20SV Wrong Token");
+            }
+            if (callValue > 0) {
+                revert("ERC20SV Non Zero Value");
+            }
+        }
+        // working with userOp.callData
+        // check if the call is to the allowed recepient and amount is not more than allowed
+        bytes calldata data;
+        {
+            uint256 offset = uint256(bytes32(_op.callData[4 + 64:4 + 96]));
```

Comment:
do you think it'd be more readable to use 68:100 and explain the calculation in comments?

Reply:
yea. think were just copying biconomies here

Reply:
https://github.com/rhinestonewtf/modulekit-examples/pull/5/commits/9e677bd6a3f981c6eff17414394f46372d935fb3